MeshGeoToolsLib/GeoMapper.cpp

```diff
@@ -352,6 +356,342 @@ void GeoMapper::advancedMapOnMesh(
+    }
+
+    if (intersections.size() == 1 && elem == beg_elem)
+    {
+        // The line segment intersects the element that contains the begin
+        // point of the line segment. Here the first sub line segment is
+        // added.
+        if (MathLib::sqrDist(beg_pnt, intersections[0]) >
+            std::numeric_limits<double>::epsilon())
+            sub_segments.emplace_back(GeoLib::LineSegment{
+                new GeoLib::Point{beg_pnt, 0},
+                new GeoLib::Point{intersections[0], 0}, true});
+    }
+
+    if (intersections.size() == 1 && elem == end_elem)
+    {
+        // The line segment intersects the element that contains the end
+        // point of the line segment. Here the last sub line segment is
+        // added.
+        if (MathLib::sqrDist(end_pnt, intersections[0]) >
+            std::numeric_limits<double>::epsilon())
+            sub_segments.emplace_back(
+                GeoLib::LineSegment{new GeoLib::Point{intersections[0], 0},
+                                    new GeoLib::Point{end_pnt, 0}, true});
+    }
+
+    if (intersections.size() == 1 && (elem != beg_elem && elem != end_elem))
+    {
+        // Since the line segment enters and leaves the element in the same
+        // point there isn't any need to insert a new sub line segment.
+        return sub_segments;
+    }
+
+    // create sub segment for the current element
+    if (intersections.size() == 2)
+    {
+        sub_segments.emplace_back(
+            GeoLib::LineSegment{new GeoLib::Point{intersections[0], 0},
+                                new GeoLib::Point{intersections[1], 0}, true});
+    }
+    return sub_segments;
+}
+
+static std::vector<GeoLib::LineSegment> mapLineSegment(
+    GeoLib::LineSegment const& segment,
+    std::vector<MeshLib::Element const*> const& surface_elements,
+    MeshLib::Element const* const beg_elem,
+    MeshLib::Element const* const end_elem)
+{
+    std::vector<GeoLib::LineSegment> sub_segments;
+    MathLib::Point3d const& beg_pnt(segment.getBeginPoint());
+    MathLib::Point3d const& end_pnt(segment.getEndPoint());
+
+    for (auto const elem : surface_elements) {
+        // compute element-segment-intersections (2d in x-y-plane)
+        std::vector<MathLib::Point3d> element_intersections(
+            computeElementSegmentIntersections(*elem, segment));
+        if (element_intersections.empty())
+            continue;
+
+        BaseLib::makeVectorUnique(element_intersections);
+
+        std::vector<GeoLib::LineSegment> sub_seg_elem(
+            createSubSegmentsForElement(element_intersections, beg_elem,
+                                        end_elem, beg_pnt, end_pnt, elem));
+        sub_segments.insert(sub_segments.end(), sub_seg_elem.begin(),
+                            sub_seg_elem.end());
+    }
+
+    // beg_elem == nullptr means there isn't any element corresponding to the
+    // beg_pnt and as a consequence the above algorithm doesn't insert a sub
+    // segment
```

Comment:
If the above loop does nothing in this case, shouldn't it then be inside the else-clause of this if statement?